### PR TITLE
refactor: Replace internal path uses by URIs

### PR DESCRIPTION
**Pull request by jirimosinger**

### CODE REVIEW

File: utils/include/utils/content_loader.h [hunk 1]
```diff
@@ -35,15 +35,26 @@ class content_loader
...
     ~content_loader() = default;
 };
 
 std::optional<std::string> load_text(const resource_location& res_loc);
 list_directory_result list_directory_files(const utils::resource::resource_location& directory_loc);
+list_directory_result list_directory_subdirs_and_symlinks(const utils::resource::resource_location& directory_loc);
 std::string filename(const utils::resource::resource_location& res_loc);
+bool file_exists(const utils::resource::resource_location& res_loc);
+bool dir_exists(const utils::resource::resource_location& res_loc);
+std::string lexically_relative(
```
Comment (jirimosinger) [inline utils/include/utils/content_loader.h:56]: I'd like to discuss where to put `lexically_*(...)` type of things and I have two proposals:

1. I can rename `content_loader` to something like `resource_operations` and leave it there. The derived classes would be renamed as well. 
2. The `lexically_*(...)` functions could be stored in `resouce_location`. `resource_location` would then also contain some logic to re-route `lexically` requests to different handlers for e.g. filesystem behavior and specific URL behaviors.

How do you @slavek-kucera feel about it?

File: parser_library/test/workspace/pathmask_test.cpp [hunk 2]
```diff
@@ -78,6 +78,15 @@ TEST(pathmask, pass)
     EXPECT_TRUE(check_path("/path/**", "\\path\\test\\"));
     EXPECT_TRUE(check_path("/path/**", "\\path\\a\\test\\"));
     EXPECT_TRUE(check_path("/path/**", "\\path\\a\\b\\test\\"));
+
+    EXPECT_TRUE(check_path("file:///C%3A/path/**/", "file:///C%3A/path/a/test/"));
```
Comment (jirimosinger) [inline parser_library/test/workspace/pathmask_test.cpp:82]: Do you @slavek-kucera think we should allow URIs with `:` as well? E.g.
```
file:///C:/path/**/"
```
instead of 
```
file:///C%3A/path/**/"
```

I would say yes as it is more user friendly but it is not a standard URI.

File: utils/include/utils/content_loader.h (hunk 1 above)
Comment (slavek-kucera) [reply utils/include/utils/content_loader.h:56]: I'd say that the `lexically_*` functions should be able to operate without any knowledge of the underlying fs.

File: parser_library/test/workspace/pathmask_test.cpp (hunk 2 above)
Comment (slavek-kucera) [reply parser_library/test/workspace/pathmask_test.cpp:82]: Is it actually an invalid URL?

More important question is that we should probably keep interpreting `C:/...` as path on windows even though it looks like an URL.

File: parser_library/src/workspaces/workspace.cpp [hunk 3]
```diff
@@ -475,66 +491,82 @@ std::regex pathmask_to_regex(std::string input)
...
+            add_diagnostic(diagnostic_s::warning_L0005(path_pattern.to_presentable(), limit));
             break;
         }
-        auto [path, canonical_path] = std::move(dirs_to_search.front());
-        dirs_to_search.pop_front();
 
-        if (!processed_canonical_paths.insert(std::move(canonical_path)).second)
+        const auto& dir = dirs_to_search.front();
+
+        if (!processed_dirs.insert(dir).second)
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace.cpp:524]: infinite loop

File: parser_library/src/workspaces/workspace.cpp [hunk 4]
```diff
@@ -475,66 +491,82 @@ std::regex pathmask_to_regex(std::string input)
...
     }
 }
 
 namespace {
+utils::resource::resource_location transform_to_resource_location(
+    std::string path, const utils::resource::resource_location& base_resource_location)
+{
+    utils::resource::resource_location rl;
+
+    if (std::filesystem::path fs_path = path; utils::path::is_absolute(fs_path))
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace.cpp:557]: what if path is url? should that be tested first?

File: parser_library/src/workspaces/workspace.h [hunk 5]
```diff
@@ -134,7 +134,7 @@ class workspace : public diagnosable_impl, public parse_lib_provider, public lsp
 private:
     constexpr static char FILENAME_PROC_GRPS[] = "proc_grps.json";
     constexpr static char FILENAME_PGM_CONF[] = "pgm_conf.json";
-    constexpr static char HLASM_PLUGIN_FOLDER[] = ".hlasmplugin";
+    constexpr static char HLASM_PLUGIN_FOLDER[] = ".hlasmplugin/";
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace.h:137]: It seems strange to me that this change was required. What's the reason behind it?

File: parser_library/test/workspace/workspace_pattern_test.cpp [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+
+    workspace ws(ws_loc, "workspace_name", file_manager, config);
+    ws.open();
+
+    EXPECT_CALL(file_manager, list_directory_files(temp_lib_loc))
+        .WillOnce(::testing::Return(list_directory_result { {}, hlasm_plugin::utils::path::list_directory_rc::done }));
+    EXPECT_CALL(file_manager, list_directory_files(temp_lib2_libs_loc))
+        .WillOnce(::testing::Return(list_directory_result { {}, hlasm_plugin::utils::path::list_directory_rc::done }));
+    EXPECT_CALL(file_manager, list_directory_files(temp_lib2_libs_subdir.first.begin()->second))
+        .WillOnce(::testing::Return(list_directory_result { {}, hlasm_plugin::utils::path::list_directory_rc::done }));
```
Comment (slavek-kucera) [inline parser_library/test/workspace/workspace_pattern_test.cpp:260]: Is this order dependent? The `list_directory_result` type contains an unordered_map, so it could cause unexpected issues...

File: utils/include/utils/resource_location.h [hunk 7]
```diff
@@ -33,9 +32,12 @@ class resource_location
     std::string get_path() const;
     std::string to_presentable(bool debug = false) const;
 
+    void to_directory();
```
Comment (slavek-kucera) [inline utils/include/utils/resource_location.h:35]: why is this needed?

File: utils/src/filesystem_content_loader.cpp [hunk 8]
```diff
@@ -58,20 +58,59 @@ std::optional<std::string> filesystem_content_loader::load_text(const resource_l
...
 std::string filesystem_content_loader::filename(const utils::resource::resource_location& res_loc) const
 {
     return utils::path::filename(std::filesystem::path(res_loc.get_path())).string();
 }
 
+bool filesystem_content_loader::file_exists(const utils::resource::resource_location& res_loc) const
+{
+    std::error_code ec;
+    return std::filesystem::exists(res_loc.get_path(), ec) && !ec && !dir_exists(res_loc);
+    // TODO use error code??
```
Comment (slavek-kucera) [inline utils/src/filesystem_content_loader.cpp:102]: it is used, right?

File: utils/src/resource_location.cpp [hunk 9]
```diff
@@ -46,19 +39,33 @@ std::string resource_location::to_presentable(bool debug) const
...
+        m_uri.append("/");
+}
+
 bool resource_location::lexically_out_of_scope() const
 {
     return m_uri == std::string_view("..") || m_uri.starts_with("../") || m_uri.starts_with("..\\");
 }
 
-resource_location resource_location::join(const resource_location& rl, std::string_view relative_path)
+void resource_location::join(std::string_view relative_path)
```
Comment (slavek-kucera) [inline utils/src/resource_location.cpp:53]: I think this should handle the case of appending the `/` in the middle when necessary.

I don't understand why this function cuts off part of the url... it should only be growing, right?

File: utils/test/resource_location_test.cpp [hunk 10]
```diff
@@ -157,10 +157,66 @@ Raw URI: aaa://user::pass@127.0.0.1:1234/path/to/resource?fileset=sources#pgm)";
...
+
+    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa:src/temp/.hlasmplugin"));
+}
+
+TEST(resource_location, join_with_root_file)
+{
+    resource_location res("aaa://src");
+    std::string rel_loc = ".hlasmplugin";
+
+    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa://.hlasmplugin"));
```
Comment (slavek-kucera) [inline utils/test/resource_location_test.cpp:195]:
```suggestion
    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa://src/.hlasmplugin"));
```

File: utils/test/resource_location_test.cpp [hunk 11]
```diff
@@ -157,10 +157,66 @@ Raw URI: aaa://user::pass@127.0.0.1:1234/path/to/resource?fileset=sources#pgm)";
...
+}
+
+
+TEST(resource_location, join_with_regular_file)
+{
+    resource_location res("aaa://src/temp");
+    std::string rel_loc = ".hlasmplugin";
+
+    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa://src/.hlasmplugin"));
```
Comment (slavek-kucera) [inline utils/test/resource_location_test.cpp:204]:
```suggestion
    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa://src/temp/.hlasmplugin"));
```

File: utils/test/resource_location_test.cpp [hunk 12]
```diff
@@ -157,10 +157,66 @@ Raw URI: aaa://user::pass@127.0.0.1:1234/path/to/resource?fileset=sources#pgm)";
...
+}
+
+TEST(resource_location, join_with_root_file_no_auth)
+{
+    resource_location res("aaa:src");
+    std::string rel_loc = ".hlasmplugin";
+
+    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa:.hlasmplugin"));
```
Comment (slavek-kucera) [inline utils/test/resource_location_test.cpp:212]:
```suggestion
    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa:src/.hlasmplugin"));
```

File: utils/test/resource_location_test.cpp [hunk 13]
```diff
@@ -157,10 +157,66 @@ Raw URI: aaa://user::pass@127.0.0.1:1234/path/to/resource?fileset=sources#pgm)";
...
+}
+
+
+TEST(resource_location, join_with_regular_file_no_auth)
+{
+    resource_location res("aaa:src/temp");
+    std::string rel_loc = ".hlasmplugin";
+
+    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa:src/.hlasmplugin"));
```
Comment (slavek-kucera) [inline utils/test/resource_location_test.cpp:221]:
```suggestion
    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa:src/temp/.hlasmplugin"));
```

File: parser_library/src/workspaces/workspace.h (hunk 5 above)
Comment (jirimosinger) [reply parser_library/src/workspaces/workspace.h:137]: The reasoning behind this is the same which goes for the unintuitive `resource_location::join()`.

Trailing slashes in the URIs denote directories whereas URIs without the trailing slash represent files -> hence the suffix here. Other solution would be to use `resource_location::to_directory()` function instead which is fine as well.

As for the `resource_location::join()` function, when following the same logic, it didn't make sense to me to do the `join()` with a file. Therefore, the file part of the URI was cut.

Anyway, I'd like to hear your thoughts on both of these topics.

I hope to find a different source but it will need to wait until tomorrow:(https://stackoverflow.com/questions/5948659/when-should-i-use-a-trailing-slash-in-my-url#:~:text=Conventionally%2C%20a%20trailing%20slash%20(%2F),the%20URL%20was%20a%20file.)

File: parser_library/src/workspaces/file_manager.h [hunk 14]
```diff
@@ -61,11 +61,18 @@ class file_manager : public virtual diagnosable
...
-    virtual list_directory_result list_directory_files(const utils::resource::resource_location& directory) = 0;
-
-    virtual bool file_exists(const std::string& file_name) = 0;
-    virtual bool lib_file_exists(const std::string& lib_path, const std::string& file_name) = 0;
+    // Returns list of all files in a directory. Returns associative array with pairs file name - file location.
+    virtual list_directory_result list_directory_files(const utils::resource::resource_location& directory) const = 0;
+
+    // Returns list of all sub directories and symbolic links. Returns associative array with pairs file path/uri - file
+    // location.
+    virtual list_directory_result list_directory_subdirs_and_symlinks(
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/workspaces/file_manager.h:69]: Basically, yes. 

I can remove this together with `list_directory_files()` and use `content_loader` directly. That would mean minor refactoring of the `workspace.cpp` as the tests would need to have the possibility to provide a mock of `content_loader` instead of the real thing. Do you want to go in this direction?

File: parser_library/src/workspaces/file_manager.h [hunk 15]
```diff
@@ -61,11 +61,18 @@ class file_manager : public virtual diagnosable
...
+        const utils::resource::resource_location& directory) const = 0;
+
+    virtual bool file_exists(const utils::resource::resource_location& file_loc) const = 0;
+    virtual bool lib_file_exists(
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/workspaces/file_manager.h]: Removed

File: parser_library/src/workspaces/workspace.cpp [hunk 16]
```diff
@@ -391,32 +393,46 @@ const processor_group& workspace::get_proc_grp(const proc_grp_id& proc_grp) cons
...
 }
 
-void replace_all(std::string& str, std::string_view what, std::string_view with)
+namespace {
+size_t preprocess_uri_with_windows_drive_letter_regex_string(const std::string& input, std::string& r)
 {
-    for (std::string::size_type p = 0; (p = str.find(what, p)) != std::string::npos; p += with.size())
-        str.replace(p, what.size(), with);
-}
+    if (std::smatch matches; std::regex_search(input, matches, std::regex("^file:///([a-zA-Z])%3[aA]/")))
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/workspaces/workspace.cpp]: Normalization added to `construct_and_analyze_lib_resource_location()` which is called prior to calling this function

File: parser_library/src/workspaces/workspace.cpp (hunk 4 above)
Comment (jirimosinger) [reply parser_library/src/workspaces/workspace.cpp:555]: I suggest leaving it in this order but adding the following chceck: `utils::path::is_directory(fs_path))`.

What do you think?

File: parser_library/src/workspaces/workspace.h (hunk 5 above)
Comment (jirimosinger) [reply parser_library/src/workspaces/workspace.h]: It turned out that this topic is not a straightforward as it would seem.

I've added some `lexically` functions and tried to mimic the behavior shown by `std::filesystem`. I've used this approach for `join()` as well so it behaves how you've suggested.

Also, `relative_reference_resolution()` based on RFC 3986 was added which behaves in a slightly similar way as was shown by the original `join` behavior (appending new path after last slash and not after last character).

File: utils/include/utils/resource_location.h (hunk 7 above)
Comment (jirimosinger) [reply utils/include/utils/resource_location.h]: The function was removed but the behavior is maintained by calling `join("")` for the discussed reason.

File: utils/src/filesystem_content_loader.cpp (hunk 8 above)
Comment (jirimosinger) [reply utils/src/filesystem_content_loader.cpp]: Well, yes. I thought that the original author of the `todo` wanted to return some specific error code when call to `exists()` fails. Anyway, removed.

File: utils/src/resource_location.cpp (hunk 9 above)
Comment (jirimosinger) [reply utils/src/resource_location.cpp]: As written above (originally based on RFC 3986 but replaced by `relative_reference_resolution()`. Join now behaves in a similar way to `std::filesystem`

File: parser_library/src/workspaces/workspace.cpp [hunk 17]
```diff
@@ -472,66 +487,85 @@ std::regex pathmask_to_regex(std::string input)
...
         if (!opts.optional_library)
-            config_diags_.push_back(diagnostic_s::error_L0001(root));
+            add_diagnostic(diagnostic_s::error_L0001(root.to_presentable()));
         return;
     }
 
+    std::regex path_validator = pathmask_to_regex(path_pattern.get_uri());
+
+    std::set<utils::resource::resource_location> processed_dirs;
+    std::deque<utils::resource::resource_location> dirs_to_search;
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace.cpp:505]: One of the reason why pair was previously used is that one entry tracked the "logical" directory nesting, while the other handled the "physical".
The logical was used to correctly handle relative directory paths, the physical was used to prevent looping. Is this behavior still preserved?

File: parser_library/test/workspace/pathmask_test.cpp [hunk 18]
```diff
@@ -93,4 +120,11 @@ TEST(pathmask, fail)
     EXPECT_FALSE(check_path("/path/a*/test/", "/path/b/test/"));
     EXPECT_FALSE(check_path("/path/*b/test/", "/path/a/test/"));
     EXPECT_FALSE(check_path("/path/a*b/test/", "/path/ba/test/"));
+
+    EXPECT_FALSE(check_path("/path/", "file:///path/"));
```
Comment (slavek-kucera) [inline parser_library/test/workspace/pathmask_test.cpp:124]: is this in principle a realistic case?

File: utils/src/path_conversions.cpp [hunk 19]
```diff
@@ -24,7 +24,7 @@
 
 namespace hlasm_plugin::utils::path {
 
-std::string uri_to_path(const std::string& uri)
+std::string uri_to_path(const std::string& uri) noexcept
```
Comment (slavek-kucera) [inline utils/src/path_conversions.cpp:27]: the `return uri;` in catch breaks noexcept - it may allocate.

File: parser_library/src/workspaces/workspace.cpp [hunk 20]
```diff
@@ -394,28 +388,33 @@ const processor_group& workspace::get_proc_grp(const proc_grp_id& proc_grp) cons
 }
 
 namespace {
+std::pair<size_t, char> match_windows_uri_with_drive(const std::string& input)
+{
+    std::smatch matches;
+    std::regex_search(input, matches, std::regex("^file:///([a-zA-Z])(?::|%3[aA])/"));
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/workspaces/workspace.cpp]: Nope

File: utils/include/utils/resource_location.h [hunk 21]
```diff
@@ -34,10 +34,11 @@ class resource_location
 
     void to_directory();
 
+    std::string lexically_relative(const resource_location& base) const;
     bool lexically_out_of_scope() const;
 
-    void join(std::string_view relative_path);
-    static resource_location join(resource_location rl, std::string_view relative_path);
+    void join(const std::string& other);
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply utils/include/utils/resource_location.h:41]: The only reason for this is that `utils::path::is_uri` needs `const std::string&` for `network::uri` which doesn't work with `std::string_view`.

File: utils/test/resource_location_test.cpp [hunk 22]
```diff
@@ -192,31 +371,53 @@ TEST(resource_location, join_with_root_file)
...
-    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa:src/.hlasmplugin"));
+    EXPECT_EQ(resource_location::join(res, rel_loc), resource_location("aaa:src/temp/.hlasmplugin"));
+}
+
+TEST(resource_location, join_prepending_slash)
+{
+    resource_location rl("aaa://src/dir/a");
+
+    rl.join("/b");
+    EXPECT_EQ(rl.get_uri(), "aaa://src/dir/a//b");
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply utils/test/resource_location_test.cpp]: Hmmm, good catch!

File: parser_library/src/workspaces/workspace.cpp (hunk 17 above)
Comment (jirimosinger) [reply parser_library/src/workspaces/workspace.cpp:500]: Alright, thanks for the insight! 

Nevertheless, I'm not really sure if this can happen anymore as the `root` and `path_pattern` are supposed to be lexically normalized before entering this function. I'll try to think about some tests for this

File: parser_library/src/workspaces/workspace.cpp (hunk 4 above)
Comment (jirimosinger) [reply parser_library/src/workspaces/workspace.cpp:547]: I think when going with URL checking first, the path "C:/dir/file" would be considered a url when it shouldn't be, right? Or what am I missing?

File: parser_library/test/workspace/pathmask_test.cpp (hunk 18 above)
Comment (jirimosinger) [reply parser_library/test/workspace/pathmask_test.cpp]: It isn't, you're right

File: parser_library/src/workspaces/workspace.cpp (hunk 4 above)
Comment (slavek-kucera) [reply parser_library/src/workspaces/workspace.cpp:547]: I see... however, there is one more case to consider - could the path constructor throw on a non-path argument?

File: parser_library/src/workspaces/workspace.cpp (hunk 17 above)
Comment (slavek-kucera) [reply parser_library/src/workspaces/workspace.cpp:500]: symlinks...

File: parser_library/src/workspaces/workspace.cpp (hunk 4 above)
Comment (jirimosinger) [reply parser_library/src/workspaces/workspace.cpp:547]: Well, according to docu this is up to implementation...

So, maybe, surround the `path` constructor with `try - catch` block. Inside the `try` block also check for `utils::path::is_directory(fs_path))`.

If the above doesn't throw and `is_directory` is `true` -> this is a filesystem path,
otherwise continue with the rest of the logic


---------------------------------------
This can potentially be enhanced by doing the above only for 1 letter schemes.

Comment (slavek-kucera) [reply parser_library/src/workspaces/workspace.cpp:547]: I think that the easiest option is for the `is_uri` to simply say false on windows for c:/ and similar "urls".

Comment (jirimosinger) [reply parser_library/src/workspaces/workspace.cpp]: Agreed and modified accordingly.

File: parser_library/src/workspaces/workspace.cpp [hunk 23]
```diff
@@ -501,37 +511,73 @@ void workspace::find_and_add_libs(
     {
         if (processed_canonical_paths.size() > limit)
         {
-            config_diags_.push_back(diagnostic_s::warning_L0005(path_pattern, limit));
+            add_diagnostic(diagnostic_s::warning_L0005(path_pattern.to_presentable(), limit));
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace.cpp:514]: I think these should be kept in `config_diags_`

File: parser_library/src/workspaces/workspace.cpp [hunk 24]
```diff
@@ -501,37 +511,73 @@ void workspace::find_and_add_libs(
...
+            add_diagnostic(diagnostic_s::error_L0001(dir.to_presentable()));
             break;
         }
+
+        for (auto& [subdir_canonical_path, subdir] : subdir_list)
+        {
+            if (processed_canonical_paths.contains(subdir_canonical_path))
+                continue;
+
+            dirs_to_search.emplace_back(subdir_canonical_path, subdir);
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace.cpp:539]:
```suggestion
            dirs_to_search.emplace_back(std::move(subdir_canonical_path), std::move(subdir));
```

File: parser_library/src/workspaces/workspace.cpp [hunk 25]
```diff
@@ -863,9 +914,8 @@ asm_option workspace::get_asm_options(const utils::resource::resource_location&
         implicit_proc_grp.update_asm_options(result);
     }
 
-    utils::resource::resource_location relative_to_location(
-        utils::path::lexically_normal(utils::path::lexically_relative(file_location.get_path(), location_.get_path()))
-            .string());
+    utils::resource::resource_location relative_to_location(file_location.lexically_relative(location_));
+    relative_to_location = utils::resource::resource_location(relative_to_location.lexically_normal());
 
     std::filesystem::path sysin_path = !pgm
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace.cpp:920]: TODO?

File: utils/src/filesystem_content_loader.cpp [hunk 26]
```diff
@@ -58,20 +58,58 @@ std::optional<std::string> filesystem_content_loader::load_text(const resource_l
...
+
+    result.second = utils::path::list_directory_subdirs_and_symlinks(path, [&result](const std::filesystem::path& p) {
+        std::error_code ec;
+        auto cp = utils::path::canonical(p, ec);
+
+        if (!ec && utils::path::is_directory(cp))
+        {
+            auto found_dir = utils::resource::resource_location(utils::path::path_to_uri(cp.string()));
+            found_dir.join(""); // Ensure that this is a directory
+            result.first[cp.string()] = std::move(found_dir);
```
Comment (slavek-kucera) [inline utils/src/filesystem_content_loader.cpp:86]: symlinks pointing to the same directory will start overwriting each other here...

Comment (jirimosinger) [reply utils/src/filesystem_content_loader.cpp]: Good point. I've switched the map parameters. I'm also open with going with a different approach if you'd like

**Review by slavek-kucera — APPROVED** [review]

File: parser_library/src/workspaces/workspace.cpp [hunk 27]
```diff
@@ -572,53 +618,69 @@ std::optional<std::string> substitute_home_directory(std::string p)
...
-            opts.extensions = lib.macro_extensions;
-        else if (!proc_groups.macro_extensions.empty())
-            opts.extensions = proc_groups.macro_extensions;
-        else if (auto macro_extensions_compatibility_list = get_macro_extensions_compatibility_list(pgm_config);
-                 !macro_extensions_compatibility_list.empty())
-        {
-            opts.extensions = macro_extensions_compatibility_list;
-            opts.extensions_from_deprecated_source = true;
-        }
+        auto lib_local_opts = get_library_local_options(lib, proc_groups, pgm_config);
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace.cpp:673]: [missing call to substitute_home_directory](https://github.com/BroadcomMFD-DevOps/che-che4z-lsp-for-hlasm/blob/ebc1ae8b1a9462e8f9856b99f3b1966a256a39b4/parser_library/src/workspaces/workspace.cpp#L591)

File: parser_library/src/workspaces/workspace.cpp [hunk 28]
```diff
@@ -388,62 +382,78 @@ const processor_group& workspace::get_proc_grp(const proc_grp_id& proc_grp) cons
...
+                if (s.starts_with("**/"))
                 {
                     if (path_started)
                     {
                         path_started = false;
-                        r.append("[^/\\\\]*[/\\\\]");
+                        r.append("[^/]*[/]");
                     }
-                    r.append("(?:[^/\\\\]+[/\\\\])*");
+                    r.append("(?:[^/]+[/])*");
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace.cpp:433]: If we allow the `//` pattern in the path, we could replace this part with simpler pattern and perhaps avoid the msvc implementation issue.
```suggestion
                    r.append("(?:.*/)?");
```

**Review by slavek-kucera — APPROVED** [review]
